### ISSUE TITLE
3.0.0: crashing with "concurrent map writes"

### ISSUE THREAD

**Issue by renchap** [imgproxy/imgproxy#755]
I updated my local imgproxy (dev environment) and its crashing when I am making concurrent request on it.

I am running imgproxy 3.0.0 on macOS 12.0.1, installed from homebrew.

Here is the full error message:

```
INFO    [2021-12-08T12:59:28+01:00] Started /unsafe/s:300:0/plain/local:///uploads/images/store/4uXBjcWw58vAedP6mhTD3-5ae60d2bdf-s1800.jpg  request_id=4VKK5g6hKnnty5XqJTdvY method=GET client_ip=127.0.0.1
INFO    [2021-12-08T12:59:28+01:00] Started /unsafe/s:300:0/plain/local:///uploads/images/store/U7VkGSrbU2tjtrBUnFiJa-45f45b9781-s1200.jpg  request_id=pddTTk4AvUx7nWtTEd5jA method=GET client_ip=127.0.0.1
fatal error: concurrent map writes
fatal error: concurrent map writes

goroutine 10 [running, locked to thread]:
runtime.throw({0x4c7a825, 0x4925be9})
	runtime/panic.go:1198 +0x71 fp=0xc0000c6fd0 sp=0xc0000c6fa0 pc=0x403ab11
runtime.mapassign_faststr(0x4aed460, 0xc0000b97a0, {0x4c73651, 0xf})
	runtime/map_faststr.go:294 +0x38b fp=0xc0000c7038 sp=0xc0000c6fd0 pc=0x401966b
github.com/imgproxy/imgproxy/v3/vips.cachedCString({0x4c73651, 0xf})
	github.com/imgproxy/imgproxy/v3/vips/cached_c_strings.go:13 +0x76 fp=0xc0000c7070 sp=0xc0000c7038 pc=0x49279f6
github.com/imgproxy/imgproxy/v3/vips.hasOperation({0x4c73651, 0xf})
	github.com/imgproxy/imgproxy/v3/vips/vips.go:123 +0x45 fp=0xc0000c70a0 sp=0xc0000c7070 pc=0x4928dc5
github.com/imgproxy/imgproxy/v3/vips.SupportsSave(0x3)
	github.com/imgproxy/imgproxy/v3/vips/vips.go:170 +0xb7 fp=0xc0000c70e8 sp=0xc0000c70a0 pc=0x49290d7
github.com/imgproxy/imgproxy/v3/processing.canSwitchFormat(0x1, 0x0, 0x3)
	github.com/imgproxy/imgproxy/v3/processing/processing.go:49 +0x2f fp=0xc0000c7100 sp=0xc0000c70e8 pc=0x49e204f
github.com/imgproxy/imgproxy/v3/processing.ProcessImage({0x4f13838, 0xc0001284e0}, 0xc000054400, 0xc00020e000)
	github.com/imgproxy/imgproxy/v3/processing/processing.go:214 +0x155 fp=0xc0000c72d0 sp=0xc0000c7100 pc=0x49e2e55
main.handleProcessing.func3({0x4f13838, 0xc0001284e0}, 0xc00015210b, 0xc0000c7668)
	github.com/imgproxy/imgproxy/v3/processing_handler.go:305 +0x74 fp=0xc0000c7338 sp=0xc0000c72d0 pc=0x49e8d14
main.handleProcessing({0xc00003c240, 0x15}, {0x4f0be98, 0xc0001c80e0}, 0xc0000d4100)
	github.com/imgproxy/imgproxy/v3/processing_handler.go:306 +0xded fp=0xc0000c78d8 sp=0xc0000c7338 pc=0x49e85ad
main.withCORS.func1({0xc00003c240, 0x15}, {0x4f0be98, 0xc0001c80e0}, 0x40139e7)
	github.com/imgproxy/imgproxy/v3/server.go:87 +0x214 fp=0xc0000c7970 sp=0xc0000c78d8 pc=0x49e9e74
github.com/imgproxy/imgproxy/v3/router.(*Router).ServeHTTP(0xc0001a2180, {0x4f0be98, 0xc0001c80e0}, 0x40698ee)
	github.com/imgproxy/imgproxy/v3/router/router.go:112 +0x5dd fp=0xc0000c7ac0 sp=0xc0000c7970 pc=0x49df57d
net/http.serverHandler.ServeHTTP({0xc0001a2360}, {0x4f0be98, 0xc0001c80e0}, 0xc0000d4000)
	net/http/server.go:2878 +0x43b fp=0xc0000c7b80 sp=0xc0000c7ac0 pc=0x429923b
net/http.(*conn).serve(0xc0003e2320, {0x4f13870, 0xc00052a180})
	net/http/server.go:1929 +0xb08 fp=0xc0000c7fb8 sp=0xc0000c7b80 pc=0x4294da8
net/http.(*Server).Serve·dwrap·82()
	net/http/server.go:3033 +0x2e fp=0xc0000c7fe0 sp=0xc0000c7fb8 pc=0x4299b8e
runtime.goexit()
	runtime/asm_amd64.s:1581 +0x1 fp=0xc0000c7fe8 sp=0xc0000c7fe0 pc=0x406bae1
created by net/http.(*Server).Serve
	net/http/server.go:3033 +0x4e8

goroutine 1 [select, 4 minutes]:
main.run()
	github.com/imgproxy/imgproxy/v3/main.go:107 +0x1a8
main.main()
	github.com/imgproxy/imgproxy/v3/main.go:126 +0x117

goroutine 35 [chan receive, 4 minutes]:
github.com/honeybadger-io/honeybadger-go.newBufferedWorker.func1()
	github.com/honeybadger-io/honeybadger-go@v0.5.0/buffered_worker.go:12 +0x47
created by github.com/honeybadger-io/honeybadger-go.newBufferedWorker
	github.com/honeybadger-io/honeybadger-go@v0.5.0/buffered_worker.go:11 +0xbe

goroutine 36 [select]:
go.opencensus.io/stats/view.(*worker).start(0xc00017a780)
	go.opencensus.io@v0.23.0/stats/view/worker.go:276 +0xb9
created by go.opencensus.io/stats/view.init.0
	go.opencensus.io@v0.23.0/stats/view/worker.go:34 +0x92

goroutine 21 [chan receive]:
main.run.func1()
	github.com/imgproxy/imgproxy/v3/main.go:83 +0x7e
created by main.run
	github.com/imgproxy/imgproxy/v3/main.go:80 +0x6b

goroutine 24 [IO wait]:
internal/poll.runtime_pollWait(0x9e446e0, 0x72)
	runtime/netpoll.go:229 +0x89
internal/poll.(*pollDesc).wait(0xc00017b000, 0x2, 0x0)
	internal/poll/fd_poll_runtime.go:84 +0x32
internal/poll.(*pollDesc).waitRead(...)
	internal/poll/fd_poll_runtime.go:89
internal/poll.(*FD).Accept(0xc00017b000)
	internal/poll/fd_unix.go:402 +0x22c
net.(*netFD).accept(0xc00017b000)
	net/fd_unix.go:173 +0x35
net.(*TCPListener).accept(0xc00000e660)
	net/tcpsock_posix.go:140 +0x28
net.(*TCPListener).Accept(0xc00000e660)
	net/tcpsock.go:262 +0x3d
golang.org/x/net/netutil.(*limitListener).Accept(0xc0001a2150)
	golang.org/x/net@v0.0.0-20211111083644-e5c967477495/netutil/listen.go:48 +0x3e
net/http.(*Server).Serve(0xc0001c8000, {0x4f0b838, 0xc0001a2150})
	net/http/server.go:3001 +0x394
main.startServer.func1()
	github.com/imgproxy/imgproxy/v3/server.go:62 +0xb9
created by main.startServer
	github.com/imgproxy/imgproxy/v3/server.go:60 +0x251

goroutine 26 [syscall, 4 minutes]:
os/signal.signal_recv()
	runtime/sigqueue.go:166 +0x28
os/signal.loop()
	os/signal/signal_unix.go:24 +0x19
created by os/signal.Notify.func1.1
	os/signal/signal.go:151 +0x2c

goroutine 9 [running, locked to thread]:
	goroutine running on other thread; stack unavailable
created by net/http.(*Server).Serve
	net/http/server.go:3033 +0x4e8

goroutine 11 [IO wait]:
internal/poll.runtime_pollWait(0x9e44510, 0x72)
	runtime/netpoll.go:229 +0x89
internal/poll.(*pollDesc).wait(0xc0000ae100, 0xc0001a2371, 0x0)
	internal/poll/fd_poll_runtime.go:84 +0x32
internal/poll.(*pollDesc).waitRead(...)
	internal/poll/fd_poll_runtime.go:89
internal/poll.(*FD).Read(0xc0000ae100, {0xc0001a2371, 0x1, 0x1})
	internal/poll/fd_unix.go:167 +0x25a
net.(*netFD).Read(0xc0000ae100, {0xc0001a2371, 0x0, 0x0})
	net/fd_posix.go:56 +0x29
net.(*conn).Read(0xc000010010, {0xc0001a2371, 0x0, 0x0})
	net/net.go:183 +0x45
net/http.(*connReader).backgroundRead(0xc0001a2360)
	net/http/server.go:672 +0x3f
created by net/http.(*connReader).startBackgroundRead
	net/http/server.go:668 +0xcf

goroutine 27 [IO wait]:
internal/poll.runtime_pollWait(0x9e445f8, 0x72)
	runtime/netpoll.go:229 +0x89
internal/poll.(*pollDesc).wait(0xc0000ae080, 0xc000618971, 0x0)
	internal/poll/fd_poll_runtime.go:84 +0x32
internal/poll.(*pollDesc).waitRead(...)
	internal/poll/fd_poll_runtime.go:89
internal/poll.(*FD).Read(0xc0000ae080, {0xc000618971, 0x1, 0x1})
	internal/poll/fd_unix.go:167 +0x25a
net.(*netFD).Read(0xc0000ae080, {0xc000618971, 0x0, 0x0})
	net/fd_posix.go:56 +0x29
net.(*conn).Read(0xc000010008, {0xc000618971, 0x0, 0x0})
	net/net.go:183 +0x45
net/http.(*connReader).backgroundRead(0xc000618960)
	net/http/server.go:672 +0x3f
created by net/http.(*connReader).startBackgroundRead
	net/http/server.go:668 +0xcf

goroutine 9 [running, locked to thread]:
runtime.throw({0x4c7a825, 0x4925be9})
	runtime/panic.go:1198 +0x71 fp=0xc0000cafd0 sp=0xc0000cafa0 pc=0x403ab11
runtime.mapassign_faststr(0x4aed460, 0xc0000b97a0, {0x4c73651, 0xf})
	runtime/map_faststr.go:294 +0x38b fp=0xc0000cb038 sp=0xc0000cafd0 pc=0x401966b
github.com/imgproxy/imgproxy/v3/vips.cachedCString({0x4c73651, 0xf})
	github.com/imgproxy/imgproxy/v3/vips/cached_c_strings.go:13 +0x76 fp=0xc0000cb070 sp=0xc0000cb038 pc=0x49279f6
github.com/imgproxy/imgproxy/v3/vips.hasOperation({0x4c73651, 0xf})
	github.com/imgproxy/imgproxy/v3/vips/vips.go:123 +0x45 fp=0xc0000cb0a0 sp=0xc0000cb070 pc=0x4928dc5
github.com/imgproxy/imgproxy/v3/vips.SupportsSave(0x3)
	github.com/imgproxy/imgproxy/v3/vips/vips.go:170 +0xb7 fp=0xc0000cb0e8 sp=0xc0000cb0a0 pc=0x49290d7
github.com/imgproxy/imgproxy/v3/processing.canSwitchFormat(0x1, 0x0, 0x3)
	github.com/imgproxy/imgproxy/v3/processing/processing.go:49 +0x2f fp=0xc0000cb100 sp=0xc0000cb0e8 pc=0x49e204f
github.com/imgproxy/imgproxy/v3/processing.ProcessImage({0x4f13838, 0xc0000b24e0}, 0xc0003dc300, 0xc00001a540)
	github.com/imgproxy/imgproxy/v3/processing/processing.go:214 +0x155 fp=0xc0000cb2d0 sp=0xc0000cb100 pc=0x49e2e55
main.handleProcessing.func3({0x4f13838, 0xc0000b24e0}, 0xc0001d400b, 0xc0000cb668)
	github.com/imgproxy/imgproxy/v3/processing_handler.go:305 +0x74 fp=0xc0000cb338 sp=0xc0000cb2d0 pc=0x49e8d14
main.handleProcessing({0xc00011a738, 0x15}, {0x4f0be98, 0xc0001fe000}, 0xc0000fc100)
	github.com/imgproxy/imgproxy/v3/processing_handler.go:306 +0xded fp=0xc0000cb8d8 sp=0xc0000cb338 pc=0x49e85ad
main.withCORS.func1({0xc00011a738, 0x15}, {0x4f0be98, 0xc0001fe000}, 0x40139e7)
	github.com/imgproxy/imgproxy/v3/server.go:87 +0x214 fp=0xc0000cb970 sp=0xc0000cb8d8 pc=0x49e9e74
github.com/imgproxy/imgproxy/v3/router.(*Router).ServeHTTP(0xc0001a2180, {0x4f0be98, 0xc0001fe000}, 0x40698ee)
	github.com/imgproxy/imgproxy/v3/router/router.go:112 +0x5dd fp=0xc0000cbac0 sp=0xc0000cb970 pc=0x49df57d
net/http.serverHandler.ServeHTTP({0xc000618960}, {0x4f0be98, 0xc0001fe000}, 0xc0000fc000)
	net/http/server.go:2878 +0x43b fp=0xc0000cbb80 sp=0xc0000cbac0 pc=0x429923b
net/http.(*conn).serve(0xc0003e20a0, {0x4f13870, 0xc00052a180})
	net/http/server.go:1929 +0xb08 fp=0xc0000cbfb8 sp=0xc0000cbb80 pc=0x4294da8
net/http.(*Server).Serve·dwrap·82()
	net/http/server.go:3033 +0x2e fp=0xc0000cbfe0 sp=0xc0000cbfb8 pc=0x4299b8e
runtime.goexit()
	runtime/asm_amd64.s:1581 +0x1 fp=0xc0000cbfe8 sp=0xc0000cbfe0 pc=0x406bae1
created by net/http.(*Server).Serve
	net/http/server.go:3033 +0x4e8
```

**Comment by renchap** [comment]
Updated to 3.1.0 and I am also able to reproduce this.

**Comment by DarthSim** [comment]
Hey @renchap! Thanks for the catch! Fixed in `v3.1.1`. I see you've created a PR to brew 👍

**Comment by renchap** [comment]
Thanks for the quick fix! I updated locally to 3.1.1, as the crash was not occurring all the time (but quite often) I will confirm in a few days that all is stable for me.

**Comment by renchap** [comment]
All is stable for me after a few days, closing the issue.

Thanks again!
